### PR TITLE
Flink backend implementation

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

This PR covers #1619.

Overall, this PR contains the following changs,

1. change `build/dist` script to support flink sql engine
2. enhance `externals/flink-sql-engine/pom.xml` to support create a shaded jar
3. simplify `externals/kyuubi-flink-sql-engine/bin/flink-sql-engine.sh`
4. introduce `FlinkSQLEngine`(flink sql engine entrypoint) and `FlinkProcessBuilder`(kyuubi server launcher)
5. add ut in kyuubi server side

After this PR, we can run the basic query e.g. `select now()` from beeline and get result, and Kyuubi Server can auto launch flink engine if there is no proper one. The Flink engine also supports other engine share levels defined in Kyuubi.

The implementation based on Flink 1.14 codebase.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FetchIterator.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.engine.flink
+
+/**
+ * Borrowed from Apache Spark, see SPARK-33655
+ */
+sealed private[engine] trait FetchIterator[A] extends Iterator[A] {
```
Comment (yanghua) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FetchIterator.scala:23]: We have already extracted the `FetchIterator` into the `kyuubi-common` for multiple engines purpose. Can we reuse it?

Comment (pan3793) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FetchIterator.scala]: Removed

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineProcessBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+          .toFile)
+        .map(_.getAbsolutePath)
+    }
+
+    flinkEngineHomeOpt.map { dir =>
+      Paths.get(dir, "bin", FLINK_ENGINE_BINARY_FILE).toAbsolutePath.toFile.getCanonicalPath
+    } getOrElse {
+      throw KyuubiSQLException("FLINK_ENGINE_HOME is not set! " +
+        "For more detail information on installing and configuring Spark, please visit " +
+        "https://kyuubi.apache.org/docs/stable/deployment/settings.html#environments")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineProcessBuilder.scala:60]: Legacy code comes from the Spark module.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineProcessBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        .flatMap(_.headOption)
+        .map(_.getAbsolutePath)
+    }
+
+    flinkHomeOpt.map { dir =>
+      dir
+    } getOrElse {
+      throw KyuubiSQLException("FLINK_ENGINE_HOME is not set! " +
+        "For more detail information on installing and configuring Spark, please visit " +
+        "https://kyuubi.apache.org/docs/stable/deployment/settings.html#environments")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineProcessBuilder.scala:160]: ditto

File: build/dist [hunk 4]
```diff
@@ -144,14 +152,17 @@ HIVE_VERSION=$("$MVN" help:evaluate -Dexpression=hive.version $@ 2>/dev/null\
...
-echo "Building Kyuubi package of version $VERSION against Spark version - $SPARK_VERSION"
+echo "Building Kyuubi package of version $VERSION against Flink $FLINK_VERSION, Spark $SPARK_VERSION"
 
-SUFFIX="-$NAME"
-if [[ "$NAME" == "none" ]]; then
-  if [[ "$SPARK_PROVIDED" == "true" ]]; then
-    SUFFIX=""
-  else
-    SUFFIX="-spark-${SPARK_VERSION:0:3}-hadoop${SPARK_HADOOP_VERSION}"
+if [[ "$NAME" != "none" ]]; then
```
Comment (pan3793) [inline build/dist:157]: `build/dist --tgz --flink-provided --spark-provided` => `apache-kyuubi-1.5.0-SNAPSHOT-bin.tgz`
`build/dist --tgz --name custom` => `apache-kyuubi-1.5.0-SNAPSHOT-bin-custom.tgz`
`build/dist --tgz --spark-provided` => `apache-kyuubi-1.5.0-SNAPSHOT-bin-flink-1.14.tgz`
`build/dist --tgz --flink-provided` => `apache-kyuubi-1.5.0-SNAPSHOT-spark-3.1-hadoop3.2.tgz`
`build/dist --tgz` => `apache-kyuubi-1.5.0-SNAPSHOT-bin-flink-1.14-spark-3.1-hadoop3.2.tgz`

Comment (pan3793) [reply build/dist:157]: Does this name pattern looks good?

File: externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/WithFlinkSQLEngine.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+import org.apache.kyuubi.KyuubiFunSuite
+import org.apache.kyuubi.config.KyuubiConf
+
+trait WithFlinkSQLEngine extends KyuubiFunSuite {
+
+  protected val flinkConfig = new Configuration()
+  protected var miniCluster: MiniCluster = _
+  protected var engine: FlinkSQLEngine = _
+  // conf will be loaded until start spark engine
```
Comment (yanghua) [inline externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/WithFlinkSQLEngine.scala:35]: `spark` keyword

File: build/dist (hunk 4 above)
Comment (yanghua) [reply build/dist:157]: currently, I am fine with the naming. But, if we introduce trino/presto, the name would be very long.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkSQLEngine.scala [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.Utils.{addShutdownHook, FLINK_ENGINE_SHUTDOWN_PRIORITY}
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.engine.flink.FlinkSQLEngine.{countDownLatch, currentEngine}
+import org.apache.kyuubi.service.Serverable
+import org.apache.kyuubi.util.SignalRegister
+
+/**
+ * A flink sql engine just like an instance of Flink SQL Engine.
```
Comment (yanghua) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkSQLEngine.scala:36]: What does this mean?

**Review by yanghua — APPROVED** [review]
LGTM

File: build/dist (hunk 4 above)
Comment (yaooqinn) [reply build/dist:157]: not good, we have a release file already

Comment (pan3793) [reply build/dist:157]: any suggestions?

Comment (pan3793) [reply build/dist:157]: > not good, we have a release file already

We only have one apache official release binary tarball `apache-kyuubi-{VERSION}-bin.tgz`

Comment (yaooqinn) [reply build/dist:157]: let's leave it as-is, we don't change the naming policy with a new engine added

Comment (pan3793) [reply build/dist]: reverted

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**